### PR TITLE
[Frontend] Add --logprobs argument to `benchmark_serving.py`

### LINKED ISSUE #8193: [Feature]: `benchmark_serving.py` should support `--logprobs`
### 🚀 The feature, motivation and pitch

The OpenAI API (and by extension vLLM's completions functionality) supports configuring the number of logprobs-per-token to return at the granularity of each request,  via the `logprobs` argument. However, `benchmarks/benchmark_serving.py` currently does not configure the `logprobs` argument when generating requests, nor does `benchmarks/benchmark_serving.py` have a `--logprobs` CLI argument. This is an issue because it is desirable to benchmark the impact of different `--logprobs` settings on vLLM performance.

So the issue is that (1) `benchmarks/benchmark_serving.py` should support a `--logprobs` argument, and (2) the value of the `--logprobs` CLI argument should configure the underlying `logprobs` argument to the completion requests generated during benchmarking.

### Alternatives

In `tests/utils.py`, the function

```
completions_with_server_args(
    prompts: List[str],
    model_name: str,
    server_cli_args: List[str],
    num_logprobs: Optional[int],
    max_wait_seconds: int = 240,
)
```

shows how to configure OpenAI API requests with the logprobs argument set. 

### Additional context

_No response_

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by afeldman-nm**
(1) benchmarks/benchmark_serving.py now supports a --logprobs CLI argument, and (2) the value of the --logprobs CLI argument configures the underlying logprobs argument to the completion requests generated during benchmarking.

```
python benchmarks/benchmark_serving.py     --backend vllm      --model JackFram/llama-160m     --dataset-name sharegpt --logprobs 4  --dataset-path ../sharegpt.json
```

FIX #8193  (*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by afeldman-nm** [comment]
CC: @comaniac @SolitaryThinker @robertgshaw2-neuralmagic @alexm-neuralmagic @njhill

File: benchmarks/benchmark_serving.py [hunk 1]
```diff
@@ -21,6 +21,10 @@
     when using tgi backend, add
         --endpoint /generate_stream
     to the end of the command above.
+
+Use --logprobs <num logprobs> to specify the number of logprobs-per-token
+to return as part of the request (or leave the argument unspecified
+to default to 1 logprob-per-token.
```
Comment (comaniac) [inline benchmarks/benchmark_serving.py:27]: We should just put this description to the `help` of `--logprobs`

File: tests/multi_step/test_correctness_llm.py [hunk 2]
```diff
@@ -57,7 +57,7 @@ def test_multi_step_llm(
                            GPU -> CPU output transfer
       num_prompts: number of example prompts under test
       num_logprobs: corresponds to the `logprobs` argument to the OpenAI
-                    completions endpoint; `None` -> no logprobs
+                    completions endpoint; `None` -> no logprobs
```
Comment (comaniac) [inline tests/multi_step/test_correctness_llm.py:60]: Miss added?

**Review by comaniac — APPROVED** [review]
LGTM. Just nits.

File: tests/multi_step/test_correctness_llm.py (hunk 2 above)
Comment (afeldman-nm) [reply tests/multi_step/test_correctness_llm.py]: Thank for this comment. It appears there is no difference between the line you are highlighting vs the corrected version. However, I assume that you are pointing out that the comment incorrectly says that `None` -> no logprobs when it should say that `None` means that 1 logprob is returned? Anyway I made that change.

File: benchmarks/benchmark_serving.py [hunk 3]
```diff
@@ -735,7 +731,8 @@ def main(args: argparse.Namespace):
         "--logprobs",
         type=int,
         default=None,
-        help="Number of logprobs-per-token to return as part of the request.",
+        help=("Number of logprobs-per-token to return as part of the request "
+              "(default 1)"),
```
Comment (comaniac) [inline benchmarks/benchmark_serving.py:735]: The `(default 1)` is a bit confusing as the default (without specifying this argument) is `None`. How the default 1 applied when I specified `--logprobs`? Or can we just let the default be `None` and always require an integer value when specifying logprobs?

Comment (afeldman-nm) [reply benchmarks/benchmark_serving.py:735]: Per our discussion, the help doc comment is now "Number of logprobs-per-token to compute & return as part of the request. If unspecified, then either (1) if beam search is disabled, no logprobs are computed & a single dummy logprob is returned for each token; or (2) if beam search is enabled 1 logprob per token is computed"

**Comment by ymoslem** [comment]
@afeldman-nm Hello! Can I also save the output of the `logprobs` themselves? Please refer to the [post here](https://discuss.vllm.ai/t/logprobs-output-from-vllm-bench-serve/1644) for details. Thanks!

**Pull request merged**